Combine psql pages into one (#3677)
timescale/docs: use-timescale/integrations/query-admin/psql.md

---
title: Connect to a Timescale Cloud service with psql 
excerpt: Install the psql client for PostgreSQL and connect to your service 
products: [cloud, mst, self_hosted]
keywords: [connect, psql]
---

# Connect with psql

> You use `psql` command line tool to interact with your $SERVICE_LONG. Most procedures in the $COMPANY documentation assume you are using `psql`.

Reviewer: ```suggestion
You use the `psql` command line tool to interact with your $SERVICE_LONG. 
```
Now we have PopSQL and SQL editor, using psql is an option, not the default.

> To use `psql` to connect to your database, you need the connection details for your PostgreSQL server. Find those in the connection string generated during service creation. For more information, see [Connecting to Timescale][about-connecting].

Reviewer: Prerequisites section for consistency.  Use the default one in the partials and add whatever is needed here.

> 1. **Check for an existing installation**

Reviewer: New section Install psql with these steps in there.

>     `psql` is sometimes installed by default, depending on your operating system and other packages you have installed over time. 

Reviewer: ```suggestion
    On many operation systems `psql` is installed by default. To use the functionality described in this documentation, best practice is to use the latest version of psql.  To check the version running on your system: 
```

    
    <Terminal>
    
    <tab label='Linux/MacOS'>
    
    ```bash
    psql --version
    ```
    
    </tab>
    
    <tab label="Windows">
    
    ```powershell
    wmic
    /output:C:\list.txt product get name, version
    ```
    
    </tab>
    

>     </Terminal>

Reviewer: Add something in here saying that if psql is already installed, proceed directly to the connect section please.

1. **Install psql**

   If there is no existing installation, take the following steps to install `psql` depending on your platform.
    
   <Tabs label="Install psql">
    

>    <Tab title="MacOS">

Reviewer: Can you make this into a couple of tabs please: `MacOS: Homebrew` and `MacOS: MacPorts`.

         ```bash
         sudo port install libpqxx
         ```
         
      1.  View the files that were installed by `libpqxx`:

         ```bash
         port contents libpqxx
         ```
         
      </Procedure>

   </Tab>
  
   <Tab title="Debian and Ubuntu">

   Install `psql` on Debian and Ubuntu with the `apt` package manager.

   <Procedure>

   1.  Make sure your `apt` repository is up to date:

       ```bash
       sudo apt-get update
       ```

   1.  Install the `postgresql-client` package:

       ```bash
       sudo apt-get install postgresql-client
       ```

   </Procedure>

   </Tab>

   <Tab title="Windows">

   `psql` is installed by default when you install PostgreSQL. This procedure uses the interactive installer provided by PostgreSQL and EnterpriseDB.

   <Procedure>

   1.  Download and run the PostgreSQL installer from [www.enterprisedb.com][windows-installer].
 
   1. In the `Select Components` dialog, check `Command Line Tools`, along with any other components you want to install, and click `Next`.

   1.  Complete the installation wizard to install the package.

   </Procedure>

   </Tab>

   </Tabs>

> 1. **Connect to your database**

Reviewer: Can you make this into a level 2 section please.

>     There are two different ways you can use `psql` to connect to your database.

Reviewer: ```suggestion
    To connect to your $SERVICE_LONG using `psql, either: 
```

    - With parameter flags:
    
        ```bash
        psql -h <HOSTNAME> -p <PORT> -U <USERNAME> -W -d <DATABASENAME>
        ```

>    - With a service URL:

Reviewer: Remove

>        - If you have configured the [SSL mode][ssl-mode] connection:

Reviewer: ```suggestion
       - Service URL with [SSL mode][ssl-mode] enabled:
```

         ```bash
         psql "postgres://tsdbadmin@<SERVICE_URL_WITH_PORT>/tsdb?sslmode=verify-full"
         ```

>        - If you haven't configured the SSL mode connection: 

Reviewer: ```suggestion
       - Service URL without ssl:
```

  
         ```bash
         psql postgres://<USERNAME>@<HOSTNAME>:<PORT>/<DATABASENAME>?sslmode=require
         ```

>        - If you want to supply your password directly within the service URL instead of being prompted for it:

Reviewer: ```suggestion
       - Service URL with password and [SSL mode][ssl-mode] enabled:
```

         psql "postgres://<USERNAME>:<PASSWORD>@<HOSTNAME>:<PORT>/<DATABASENAME>?sslmode=require"
         ```

## Useful psql commands

When you start using `psql`, these are the commands you are likely to use most
frequently:

|Command|Description|
|-|-|
|`\c <DB_NAME>`|Connect to a new database|
|`\d <TABLE_NAME>`|Show the details of a table|
|`\df`|List functions in the current database|
|`\df+`|List all functions with more details|
|`\di`|List all indexes from all tables|
|`\dn`|List all schemas in the current database|
|`\dt`|List available tables|
|`\du`|List PostgreSQL database roles|
|`\dv`|List views in current schema|
|`\dv+`|List all views with more details|
|`\dx`|Show all installed extensions|
|`ef <FUNCTION_NAME>`|Edit a function|
|`\h`|Show help on syntax of SQL commands|
|`\l`|List available databases|
|`\password <USERNAME>`|Change the password for the user|
|`\q`|Quit `psql`|
|`\set`|Show system variables list|
|`\timing`|Show how long a query took to execute|
|`\x`|Show expanded query results|
|`\?`|List all `psql` slash commands|

For more on `psql` commands, see the [Timescale psql cheat sheet][psql-cheat-sheet] and [psql documentation][psql-docs].

> ### Save query results to a file

Reviewer: ```suggestion
## Save query results to a file
```

When you run queries in `psql`, the results are shown in the console by default.
If you are running queries that have a lot of results, you might like to save
the results into a comma-separated `.csv` file instead. You can do this using
the `COPY` command. For example:

```sql
\copy (SELECT * FROM ...) TO '/tmp/output.csv' (format CSV);
```

This command sends the results of the query to a new file called `output.csv` in
the `/tmp/` directory. You can open the file using any spreadsheet program.

> ### Edit queries in a text editor

Reviewer: ```suggestion
## Edit queries in a text editor
```

---
title: Connect to a Timescale Cloud service with psql 
excerpt: Install the psql client for PostgreSQL and connect to your service 
products: [cloud, mst, self_hosted]
keywords: [connect, psql]
---

import IntegrationPrereqs from "versionContent/_partials/_integration-prereqs.mdx";

# Connect with psql

You use the `psql` command line tool to interact with your $SERVICE_LONG.

## Prerequisites

<IntegrationPrereqs />

> 

Reviewer: Can you add a line here saying you can also use psql to connect to a self-hosted deployment of TimescaleDB.

<IntegrationPrereqs />

## Check for an existing installation

On many operating systems, `psql` is installed by default. To use the functionality described in this page, best practice is to use the latest version of `psql`. To check the version running on your system:

<Terminal>
    
<tab label='Linux/MacOS'>
    

```bash
psql --version
```
    
</tab>
    
<tab label="Windows">

    
```powershell
wmic
/output:C:\list.txt product get name, version
```

</tab>
    
</Terminal>

If you already have the latest version of `psql` installed, proceed to the [Connect to your $SERVICE_SHORT][connect-database] section. 

## Install psql

> If there is no existing installation, take the following steps to install `psql` depending on your platform.

Reviewer: ```suggestion
If there is no existing installation, take the following steps to install `psql`:
```

    
<Tabs label="Install psql">
    
<Tab title="MacOS Homebrew">

Install using Homebrew. `libpqxx` is the official C++ client API for PostgreSQL.

<Procedure>

1. Install Homebrew, if you don't already have it:

    ```bash
    /bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
    ```

    For more information about Homebrew, including installation instructions, see the [Homebrew documentation][homebrew].
 
1. Make sure your Homebrew repository is up to date:

    ```bash
    brew doctor
    brew update
    ```

1. Install `psql`:

    ```bash
    brew install libpq
    ```

1. Update your path to include the `psql` tool:

    ```bash
    brew link --force libpq
    ```

On Intel chips, the symbolic link is added to `/usr/local/bin`. On Apple Silicon, the symbolic link is added to `/opt/homebrew/bin`.

</Procedure>

</Tab>

<Tab title="MacOS MacPorts">

Install using MacPorts. `libpqxx` is the official C++ client API for PostgreSQL.

<Procedure>

> 1. Install MacPorts by downloading and running the package installer.

Reviewer: ```suggestion
1. [Install MacPorts][macports] by downloading and running the package installer.
```

>     For more information about MacPorts, including installation instructions, see the [MacPorts documentation][macports].

Reviewer: remove please.

</Tab>

<Tab title="Windows">

`psql` is installed by default when you install PostgreSQL. This procedure uses the interactive installer provided by PostgreSQL and EnterpriseDB.

<Procedure>

1.  Download and run the PostgreSQL installer from [www.enterprisedb.com][windows-installer].
 
1. In the `Select Components` dialog, check `Command Line Tools`, along with any other components you want to install, and click `Next`.

1.  Complete the installation wizard to install the package.

</Procedure>

</Tab>

</Tabs>

## Connect to your $SERVICE_SHORT

To use `psql` to connect to your $SERVICE_SHORT, you need the connection details. See [Find your connection details][connection-info].

Connect to your $SERVICE_SHORT with either:

- The parameter flags:
    
   ```bash
   psql -h <HOSTNAME> -p <PORT> -U <USERNAME> -W -d <DATABASENAME>
   ```

- The $SERVICE_SHORT URL with the [SSL mode][ssl-mode] enabled:

   ```bash
   psql "postgres://<USERNAME>@<HOSTNAME>:<PORT>/<DATABASENAME>?sslmode=verify-full"
   ```

- The $SERVICE_SHORT URL with the password and the SSL mode enabled:

   ```bash
   psql "postgres://<USERNAME>:<PASSWORD>@<HOSTNAME>:<PORT>/<DATABASENAME>?sslmode=require"
   ```

> - The $SERVICE_SHORT URL without SSL:

Reviewer: This is with SSL. sslmode=require.
